libdnf/repo/repo_downloader.hpp

```diff
@@ -26,6 +26,7 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "libdnf/base/base_weak.hpp"
 #include "libdnf/common/exception.hpp"
 #include "libdnf/repo/config_repo.hpp"
+#include "libdnf/repo/repo.hpp"
```

Comment:
What I don't like about this is that it creates a virtual cycle in the class hierarchy. `Repo` uses `RepoDownloader` and now you've added the reverse dependency of `RepoDownloader` using `Repo`(`::LoadFlags`).

I know it works right now, but it could cause issues in the future.

To maintain a uni-directional hierarchy the `LoadFlags` should be split out of the `Repo` class and into its own header file I think. I've been meaning to go in this direction.

Reply:
I wouldn't call it laziness, you just don't download the data you don't need. But on the next run, with a different command, more data can be needed and it just won't be there. I don't see how we can do this unless we check what metadata are actually downloaded and load anything else that's needed on-demand.

Another "issue" would be automatic downloading metadata in the background, what would that download? All of them?

Reply:
> Another "issue" would be automatic downloading metadata in the background, what would that download? All of them?

All metadata for dependency resolution and applying updateinfo should be downloaded by default.
This means: primary, modulemd, updateinfo, maybe also comps if we're going to support upgrading groups.
In other words: filelists and others (== changelogs) are not required in most cases.

Reply:
Ok. I guess that's a detail, the list of metadata to download for automatic can be either hardcoded or even eventually configurable.

You didn't address my first point, which is the real issue here AFAICS. Seems that'll be harder to tackle, how about removing the "lazy" loading from the PR and keep the downloaded metadata set the same (and just separate the `LoadFlags` into its own header). I'd merge that and you can post another PR for the loading if/when you want to sort out the "additional" downloading issue.

We'd prefer PRs to be merged instead of hanging around indefinitely :grin:

Reply:
I'll make the changes incl. `LoadFlags` refactoring as soon as time allows.
Thanks for your feedback.